src/mlpack/core/optimizers/nestrov/nag_impl.hpp

```diff
@@ -0,0 +1,127 @@
+                                   const size_t maxIterations,
+                                   const double tolerance,
+                                   const double momentum,
+                                   const bool shuffle) :
+    function(function),
+    stepSize(stepSize),
+    maxIterations(maxIterations),
+    tolerance(tolerance),
+    momentum(momentum),
+    shuffle(shuffle)
+{ /* Nothing to do. */ }
+
+//! Optimize the function (minimize).
+template<typename DecomposableFunctionType>
+double NAG<DecomposableFunctionType>::Optimize(arma::mat& iterate)
+{
+  // Find the number of functions to use.
+  const size_t numFunctions = function.NumFunctions();
+
+  // This is used only if shuffle is true.
+  arma::Col<size_t> visitationOrder;
+  if (shuffle)
+    visitationOrder = arma::shuffle(arma::linspace<arma::Col<size_t>>(0,
+        (numFunctions - 1), numFunctions));
+
+  // To keep track of where we are and how things are going.
+  size_t currentFunction = 0;
+  double overallObjective = 0;
+  double lastObjective = DBL_MAX;
+
+  // Calculate the first objective function.
+  for (size_t i = 0; i < numFunctions; ++i)
+    overallObjective += function.Evaluate(iterate, i);
+
+  // Now iterate!
+  arma::mat iterate_temp = iterate;
+  arma::mat gradient(iterate.n_rows, iterate.n_cols);
+  arma::mat velocity = arma::zeros<arma::mat>(iterate.n_rows, iterate.n_cols);
+  for (size_t i = 1; i != maxIterations; ++i, ++currentFunction)
+  {
+    // Is this iteration the start of a sequence?
+    if ((currentFunction % numFunctions) == 0)
+    {
+      // Output current objective function.
+      Log::Info << "NAG: iteration " << i << ", objective " << overallObjective
+          << "." << std::endl;
+
+      if (std::isnan(overallObjective) || std::isinf(overallObjective))
+      {
+        Log::Warn << "NAG: converged to " << overallObjective << "; terminating"
+            << " with failure.  Try a smaller step size?" << std::endl;
+        return overallObjective;
+      }
+
+      if (std::abs(lastObjective - overallObjective) < tolerance)
+      {
+        Log::Info << "NAG: minimized within tolerance " << tolerance << "; "
+            << "terminating optimization." << std::endl;
+        return overallObjective;
+      }
+
+      // Reset the counter variables.
+      lastObjective = overallObjective;
+      overallObjective = 0;
+      currentFunction = 0;
+      if (shuffle) // Determine order of visitation.
+        visitationOrder = arma::shuffle(visitationOrder);
+    }
+
+    // Evaluate the gradient at the modified iteration
+    if (shuffle)
+      function.Gradient(iterate + momentum * velocity, visitationOrder[currentFunction], gradient);
+    else
+      function.Gradient(iterate + momentum * velocity, currentFunction, gradient);
+
+    //iterate += momentum * velocity;
+    velocity = velocity*momentum - stepSize*gradient;
+
+    // And update the iterate.
+    iterate = iterate + velocity;
```

Comment:
@zoq I am thinking we could have  2 base classes 1. sgd(arun) 2.nestrov. I want to implement nestrov as a base class where we all the accelerated variants of optimisers can be implemented with nestrov being the base the class and all other being policies. This is similar to how caffe does things(though) they use inheritance. Also why do we prefer policy based design pattern over inheritance(they are pretty similar).